cameo/flux_analysis/structural.py

```diff
@@ -16,25 +16,118 @@
+    ----------
+    matrix : ndarray
+        A should be at most 2-D.  A 1-D array with length k will be treated
+        as a 2-D with shape (1, k)
+    atol : float
+        The absolute tolerance for a zero singular value.  Singular values
+        smaller than `atol` are considered to be zero.
+    rtol : float
+        The relative tolerance.  Singular values less than rtol*smax are
+        considered to be zero, where smax is the largest singular value.
+
+    If both `atol` and `rtol` are positive, the combined tolerance is the
+    maximum of the two; that is::
+        tol = max(atol, rtol * smax)
+    Singular values smaller than `tol` are considered to be zero.
+
+    Return value
+    ------------
+    ns : ndarray
+        If `A` is an array with shape (m, k), then `ns` will be an array
+        with shape (k, n), where n is the estimated dimension of the
+        nullspace of `A`.  The columns of `ns` are a basis for the
+        nullspace; each element in numpy.dot(A, ns) will be approximately
+        zero.
+    """
+
+    matrix = np.atleast_2d(matrix)
+    u, s, vh = svd(matrix)
+    tol = max(atol, rtol * s[0])
+    nnz = (s >= tol).sum()
+    ns = vh[nnz:].conj().T
+    return ns
+
+
+def find_blocked_reactions_nullspace(model, ns=None, tol=1e-10):
+    """Identify reactions that can't carry flux based on the nullspace of the stoichiometric matrix.
+    A blocked reaction will correspond to an all-zero row in N(S)."""
+    if ns is None:
+        ns = nullspace(model.S)
+    mask = (np.abs(ns) <= tol).all(1)
+    blocked = [reac for reac, b in zip(model.reactions, mask) if b is True]
+    return blocked
+
+
+def find_coupled_reactions_nullspace(model, ns=None, tol=1e-10):
+    """
+    Find groups of reactions whose fluxes are forced to be multiples of each other.
+    """
+    if ns is None:
+        ns = nullspace(model.S)
+    blocked = set(find_blocked_reactions_nullspace(model, ns, tol))
+    blocked_mask = np.array([True if r in blocked else False for r in model.reactions])
+    non_blocked_ns = ns[~blocked_mask]
+    non_blocked_reactions = np.array(list(model.reactions))[~blocked_mask]
+
+    # Calculate correlation coefficients (perfect correlation means that one is a scalar multiple of the other)
+    corr_mat = np.corrcoef(non_blocked_ns)
+    dist_mat = 1 - np.abs(corr_mat)
+
+    # Perform a linkage clustering
+    link = linkage(dist_mat, method="complete")
+
+    # Filter the linkage results (only distance = 0)
+    good_link = [a for a in link if a[2] < tol]
+
+    # Aggregate groups
+    group_dict = {i: {i} for i in range(len(non_blocked_reactions))}
+    for i, n in enumerate(range(len(non_blocked_reactions), len(non_blocked_reactions) + len(good_link))):
+        a = good_link[i]
+        group_dict[n] = group_dict[a[0]] | group_dict[a[1]]
+        del group_dict[a[0]]
+        del group_dict[a[1]]
+
+    groups = [set(non_blocked_reactions[np.array(list(v))]) for v in group_dict.values()]
```

Comment:
You can also index with a list directly into an array rather than creating an array.